### PR TITLE
UNR-2363 Use accessors to get Spatial Networking flag from settings

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Uses accessors for bSpatialNetworking instead of the class member.
Add a test which checks we get the correct setting for Spatial networking early in the startup process.
Depends on PR : https://github.com/improbableio/UnrealEngine/pull/287
#### Tests
Added two tests : 
- SpatialActivationReport, which logs the state of the SpatialActivation flag
- SpatialActivationOverride, which launches a sub-process running SpatialActivationReport with various combination of settings and reads the log to determine if we got the flag state we expect.

### CODE REVIEW

**Comment by m-samiec** [comment]
Can you remove the `draft` tag. And please provide a direct URL link to the engine PR.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+#include "CoreMinimal.h"
+
+#include "TestDefinitions.h"
+#include "Tests/AutomationCommon.h"
+#include "Runtime/EngineSettings/Public/EngineSettings.h"
+
+static const bool s_earliestFlag =  GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking();
+
+FString const s_earliestFlagReport = "Spatial activation Flag [Earliest]:";
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]: Please capitalize the first letter of variables. And the only hungarian notation we use is for booleans.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "CoreMinimal.h"
+
+#include "TestDefinitions.h"
+#include "Tests/AutomationCommon.h"
+#include "Runtime/EngineSettings/Public/EngineSettings.h"
+
+static const bool s_earliestFlag =  GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]: Is this safe? Might the settings not be initialized at this point?

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]: Please wrap all strings in the TEXT() macro

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp (hunk 2 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]: I test this case, and apparently we get the settings we expect.
The test plugin is set to be loaded before the SpatialGDKServices.
I admit this is a bit flaky, and heavily dependent on ordering of many other parts, but this is like what the DeploymentManager is doing at the moment.

File: SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/CookAndGenerateSchemaCommandlet.cpp [hunk 3]
```diff
@@ -68,7 +68,7 @@ UCookAndGenerateSchemaCommandlet::UCookAndGenerateSchemaCommandlet()
 int32 UCookAndGenerateSchemaCommandlet::Main(const FString& CmdLineParams)
 {
 	// Force spatial networking
-	GetMutableDefault<UGeneralProjectSettings>()->bSpatialNetworking = true;
+	GetMutableDefault<UGeneralProjectSettings>()->SetUseSpatialNetworking(true);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/CookAndGenerateSchemaCommandlet.cpp]: Driveby, could you wrap this in the #if ENGINE_MINOR_VERSION tags, like above?
cc: @mironec

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+#include "TestDefinitions.h"
+#include "Tests/AutomationCommon.h"
+#include "Runtime/EngineSettings/Public/EngineSettings.h"
+
+namespace
+{
+	const bool bEarliestFlag = GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking();
+
+	FString const EarliestFlagReport = TEXT("Spatial activation Flag [Earliest]:");
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]:
```suggestion
	const FString EarliestFlagReport = TEXT("Spatial activation Flag [Earliest]:");
```

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	UBoolProperty* SpatialFlagProperty = Cast<UBoolProperty>(UGeneralProjectSettings::StaticClass()->FindPropertyByName("bSpatialNetworking"));
+	TestNotNull("Property existence", SpatialFlagProperty);
+
+	UGeneralProjectSettings const* ProjectSettings = GetDefault<UGeneralProjectSettings>();
+	TestNotNull("Settings existence", ProjectSettings);
+
+	// Bypass C++ function to get the property saved to settings.
+	bool SettingsPropertyValue = SpatialFlagProperty->GetPropertyValue(SpatialFlagProperty->ContainerPtrToValuePtr<void const*>(ProjectSettings));
+
+	UE_LOG(LogTemp, Display, TEXT("%s %i"), EarliestFlagReport.GetCharArray().GetData(), bEarliestFlag);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]:
```suggestion
	UE_LOG(LogTemp, Display, TEXT("%s %i"), *EarliestFlagReport, bEarliestFlag);
```

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+GDK_TEST(Core, UGeneralProjectSettings, SpatialActivationReport)
+{
+	UBoolProperty* SpatialFlagProperty = Cast<UBoolProperty>(UGeneralProjectSettings::StaticClass()->FindPropertyByName("bSpatialNetworking"));
+	TestNotNull("Property existence", SpatialFlagProperty);
+
+	UGeneralProjectSettings const* ProjectSettings = GetDefault<UGeneralProjectSettings>();
+	TestNotNull("Settings existence", ProjectSettings);
+
+	// Bypass C++ function to get the property saved to settings.
+	bool SettingsPropertyValue = SpatialFlagProperty->GetPropertyValue(SpatialFlagProperty->ContainerPtrToValuePtr<void const*>(ProjectSettings));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]:
```suggestion
	bool bSettingsPropertyValue = SpatialFlagProperty->GetPropertyValue(SpatialFlagProperty->ContainerPtrToValuePtr<void const*>(ProjectSettings));
```

File: SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/CookAndGenerateSchemaCommandlet.cpp (hunk 3 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/CookAndGenerateSchemaCommandlet.cpp]: Ah yes, I must have missed this, it should be wrapped in #if ENGINE_MINOR_VERSION <= 22
since this is no longer required in 4.23.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+namespace
+{
+	struct ReportedFlags
+	{
+		bool bEarliestFlag;
+		bool bSettingsFlag;
+		bool bCurrentFlag;
+	};
+
+	ReportedFlags RunSubProcessAndExtractFlags(FAutomationTestBase& test, FString const& commandLineArgs)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]:
```suggestion
	ReportedFlags RunSubProcessAndExtractFlags(FAutomationTestBase& Test, const FString& commandLineArgs)
```

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+		int32 ReturnCode = 1;
+		FString StdOut;
+		FString StdErr;
+
+		FPlatformProcess::ExecProcess(TEXT("UE4Editor"), commandLineArgs.GetCharArray().GetData(), &ReturnCode, &StdOut, &StdErr);
+
+		test.TestTrue("Sucessful run", ReturnCode == 0);
+
+		auto extractFlag = [&](FString const& pattern, bool& flag)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]:
```suggestion
		auto extractFlag = [&](FString const& Pattern, bool& bFlag)
```

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+		return Flags;
+	}
+}
+
+GDK_TEST(Core, UGeneralProjectSettings, SpatialActivationOverride)
+{
+	FString ProjectPath = FPaths::GetProjectFilePath();
+	FString CommandLineArgs = ProjectPath;
+	CommandLineArgs.Append(" -ExecCmds=\"Automation RunTests SpatialGDK.Core.UGeneralProjectSettings.SpatialActivationReport; Quit\"");
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]: These should all be wrapped in TEXT() macro

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+	void* SpatialFlagPtr = SpatialFlagProperty->ContainerPtrToValuePtr<void const*>(ProjectSettings);
+
+	bool SavedFlagValue = SpatialFlagProperty->GetPropertyValue(SpatialFlagPtr);
+
+	{
+		ProjectSettings->SetUseSpatialNetworking(false);
+		ProjectSettings->UpdateSinglePropertyInConfigFile(SpatialFlagProperty, ProjectSettings->GetDefaultConfigFilename());
+
+		auto flags = RunSubProcessAndExtractFlags(*this, CommandLineArgs);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]: Generally don't use auto except for iterators and function signatures/lambdas

File: SpatialGDK/SpatialGDK.uplugin [hunk 11]
```diff
@@ -48,7 +48,7 @@
     {
       "Name": "SpatialGDKTests",
       "Type": "Editor",
-      "LoadingPhase": "Default",
+      "LoadingPhase": "PreLoadingScreen",
```
Comment (m-samiec) [inline SpatialGDK/SpatialGDK.uplugin]: Why this change?

Comment (ImprobableNic) [reply SpatialGDK/SpatialGDK.uplugin]: Some plugins are loaded as PreDefault. I'm capturing the setting available when the test plugin is loaded through a gloabl variable(although I could do it in the module initialization function, to be sure), and I'm checking that its value is the expected one in the tests.
If the test pass that way, it should ensure that any plugin loaded from the PreDefault stage has access to the correct value of the flag.

Comment (m-samiec) [reply SpatialGDK/SpatialGDK.uplugin]: I wonder if this will have impacts on future tests we want to write. @aleximprobable do you have an opinion?

Comment (aleximprobable) [reply SpatialGDK/SpatialGDK.uplugin]: No. However, I think it should be fine as long as existing tests pass. And if it breaks future tests - we can revisit this decision.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+	{
+		ReportedFlags Flags;
+
+		int32 ReturnCode = 1;
+		FString StdOut;
+		FString StdErr;
+
+		FPlatformProcess::ExecProcess(TEXT("UE4Editor"), *CommandLineArgs, &ReturnCode, &StdOut, &StdErr);
+
+		Test.TestTrue("Sucessful run", ReturnCode == 0);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlags.cpp]:
```suggestion
		Test.TestTrue("Successful run", ReturnCode == 0);
```

**Comment by m-samiec** [comment]
Could you also add a note to the release note to the CHANGELOG.md in the bugfix section that details the fix.
